File: backend/app/schemas/subscriptions.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class PlanRead(BaseModel):
    key: str
    name: str
    description: str
    monthly_credits: Decimal
    max_source_clips: int
    max_target_duration_seconds: int
    max_members: int
    max_tier: int
    checkout_available: bool
    current: bool = False


class SubscriptionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    workspace_id: UUID
    provider: str
    plan_key: str
    status: str
    current_period_start: datetime | None
    current_period_end: datetime | None
    cancel_at_period_end: bool
    last_payment_failed_at: datetime | None
    updated_at: datetime


class BillingOverviewRead(BaseModel):
    workspace_id: UUID
    plan: PlanRead
    subscription: SubscriptionRead | None
    balance_credits: Decimal
    reserved_credits: Decimal
    available_credits: Decimal
    portal_available: bool


class CheckoutCreate(BaseModel):
    plan_key: str = Field(min_length=2, max_length=64, pattern=r"^[a-z][a-z0-9_-]+$")


class HostedBillingSessionRead(BaseModel):
    url: str
    expires_at: datetime | None = None


class WebhookAccepted(BaseModel):
    received: bool = True
    duplicate: bool = False
    status: str

File: frontend/lib/subscriptions.ts
import { ApiError, getAccessToken } from "@/lib/api";

const API_URL = (process.env.NEXT_PUBLIC_DIRECTOR_API_URL ?? "http://localhost:8000/api/v1").replace(/\/$/, "");

export interface BillingPlan {
  key: string;
  name: string;
  description: string;
  monthly_credits: string;
  max_source_clips: number;
  max_target_duration_seconds: number;
  max_members: number;
  max_tier: number;
  checkout_available: boolean;
  current: boolean;
}

export interface WorkspaceSubscription {
  workspace_id: string;
  provider: string;
  plan_key: string;
  status: string;
  current_period_start: string | null;
  current_period_end: string | null;
  cancel_at_period_end: boolean;
  last_payment_failed_at: string | null;
  updated_at: string;
}

export interface SubscriptionOverview {
  workspace_id: string;
  plan: BillingPlan;
  subscription: WorkspaceSubscription | null;
  balance_credits: string;
  reserved_credits: string;
  available_credits: string;
  portal_available: boolean;
}

interface HostedBillingSession {
  url: string;
  expires_at: string | null;
}

async function subscriptionRequest<T>(path: string, init?: RequestInit): Promise<T> {
  const token = getAccessToken();
  const response = await fetch(`${API_URL}${path}`, {
    ...init,
    headers: {
      Accept: "application/json",
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
      ...(init?.body ? { "Content-Type": "application/json" } : {}),
      ...init?.headers,
    },
    cache: "no-store",
  });
  if (!response.ok) {
    let message = `Request failed with status ${response.status}`;
    try {
      const payload = (await response.json()) as { detail?: string };
      if (payload.detail) message = payload.detail;
    } catch {
      // Preserve the status-based fallback.
    }
    throw new ApiError(message, response.status);
  }
  return (await response.json()) as T;
}

export function listBillingPlans(): Promise<BillingPlan[]> {
  return subscriptionRequest<BillingPlan[]>("/billing/plans");
}

export function getSubscriptionOverview(workspaceId: string): Promise<SubscriptionOverview> {
  return subscriptionRequest<SubscriptionOverview>(`/workspaces/${workspaceId}/subscription`);
}

export function createSubscriptionCheckout(
  workspaceId: string,
  planKey: string,
): Promise<HostedBillingSession> {
  return subscriptionRequest<HostedBillingSession>(
    `/workspaces/${workspaceId}/subscription/checkout`,
    { method: "POST", body: JSON.stringify({ plan_key: planKey }) },
  );
}

export function createSubscriptionPortal(workspaceId: string): Promise<HostedBillingSession> {
  return subscriptionRequest<HostedBillingSession>(
    `/workspaces/${workspaceId}/subscription/portal`,
    { method: "POST" },
  );
}
